### PR TITLE
webgpu: support reverse operator

### PR DESCRIPTION
**Pull request by xhcao**
To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

- - -
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6973)

### CODE REVIEW

**Comment by xhcao** [comment]
@hujiajie @qjia7 PTAL

File: tfjs-backend-webgpu/src/reverse_webgpu.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        this.dispatchLayout, this.outputShape, this.workgroupSize);
+    this.uniforms = ` axis : vec4<i32>,`;
+    this.shaderKey = 'reverse';
+  }
+
+  getUserCode(): string {
+    const inputCoordsSnippet = `
+      fn getReverseInputCoords(coords : vec4<i32>) -> vec4<i32> {
+        var inputCoords: vec4<i32>;
+        if ((uniforms.axis[0] == 1) && (uniforms.xShape[0] != 1)) {
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/reverse_webgpu.ts:44]: From performance point, I'd prefer you don't make axis as a uniform. And  meanwhile squeeze `x` so that there is no `1` in shape. In this case, we can avoid all `if` here?

Comment (qjia7) [reply tfjs-backend-webgpu/src/reverse_webgpu.ts:44]: Or we can still keep axis as a uniform, but update the value in it so that we can always use `inputCoords[i] = uniforms.xShape[1] - coords[1] - 1`?

Comment (xhcao) [reply tfjs-backend-webgpu/src/reverse_webgpu.ts:44]: When we launching a kernel, these values of `uniforms.axis[0...3] == 1` and `uniforms.xShape[0...3] != 1` are the same for all threads, all threads enter into either `if block` or `else block`. There is no situation that some threads enter `if block`, and others enter `else block`. So I think there is no performance issue. Do you think so?

Comment (qjia7) [reply tfjs-backend-webgpu/src/reverse_webgpu.ts:44]: Thanks for the explanation. Please add this comment into the code in case someone revisits this code in future.

**Review by qjia7 — APPROVED** [review]
LGTM, thanks!

File: tfjs-backend-webgpu/src/reverse_webgpu.ts (hunk 1 above)
Comment (xhcao) [reply tfjs-backend-webgpu/src/reverse_webgpu.ts:46]: Done

File: tfjs-backend-webgpu/src/kernels/Reverse.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  const xShape = x.shape;
+  const xShape4D: [number, number, number, number] = [1, 1, 1, 1];
+  xShape.forEach((d, i) => {
+    const index = i + 4 - xRank;
+    xShape4D[index] = d;
+  });
+
+  const $dims = util.parseAxisParam(dims, x.shape);
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/kernels/Reverse.ts:45]: Why is the variable name starting with $?

File: tfjs-backend-webgpu/src/reverse_webgpu.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        if ((uniforms.axis[2] == 1) && (uniforms.xShape[2] != 1)) {
+          inputCoords[2] = uniforms.xShape[2] - coords[2] - 1;
+        } else {
+          inputCoords[2] = coords[2];
+        }
+        if ((uniforms.axis[3] == 1) && (uniforms.xShape[3] != 1)) {
+          inputCoords[3] = uniforms.xShape[3] - coords[3] - 1;
+        } else {
+          inputCoords[3] = coords[3];
+        }
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/reverse_webgpu.ts:65]: Can we simplify the impl to modify coords directly?

File: tfjs-backend-webgpu/src/reverse_webgpu.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+        return inputCoords;
+      }
+    `;
+    const userCode = `
+      ${inputCoordsSnippet}
+      ${main('index')} {
+        if (index < uniforms.size) {
+          let coords = getCoordsFromIndex(index);
+          let inputCoords = getReverseInputCoords(coords);
```
Comment (gyagp) [inline tfjs-backend-webgpu/src/reverse_webgpu.ts:75]: inputCoords -> reverseCoords?

File: tfjs-backend-webgpu/src/reverse_webgpu.ts (hunk 3 above)
Comment (xhcao) [reply tfjs-backend-webgpu/src/reverse_webgpu.ts:57]: I removed `else block` here, but had no good idea how to simplify other codes.

File: tfjs-backend-webgpu/src/reverse_webgpu.ts (hunk 4 above)
Comment (xhcao) [reply tfjs-backend-webgpu/src/reverse_webgpu.ts]: Done, also changed `getReverseInputCoords` to `getReverseCoords`, and changed `inputCoords` to `reverseCoords` in `getReverseCoords` function.

File: tfjs-backend-webgpu/src/kernels/Reverse.ts (hunk 2 above)
Comment (xhcao) [reply tfjs-backend-webgpu/src/kernels/Reverse.ts]: I had changed `$dims` to `axes`. Using `$` referred other codes.
$ grep parseAxisParam -rsn .
./kernels/ArgMax.ts:32:  let axes = util.parseAxisParam(axis, x.shape);
./kernels/ArgMin.ts:32:  let axes = util.parseAxisParam(axis, x.shape);
./kernels/Concat.ts:31:  const $axis = util.parseAxisParam(axis, inputs[0].shape)[0];
./kernels/GatherV2.ts:36:  const parsedAxis = util.parseAxisParam(axis, x.shape)[0];
./kernels/Softmax.ts:36:  const axes = util.parseAxisParam([dim], logits.shape);
./kernels/SplitV.ts:30:  const $axis = util.parseAxisParam(axis, x.shape)[0];
./kernel_utils/reduce.ts:36:  const origAxes = util.parseAxisParam(axis, x.shape);

**Comment by xhcao** [comment]
@gyagp PTAL

**Review by gyagp — APPROVED** [review]
LGTM
